File: languages/go/client/identity_v2.go
package client

import (
	"context"
	"errors"

	"github.com/bloock/bloock-sdk-go/v2/entity/authenticity"
	"github.com/bloock/bloock-sdk-go/v2/entity/identityV2"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
	"github.com/bloock/bloock-sdk-go/v2/internal/config"
)

// IdentityClient represents a client for interacting with the [Bloock Identity service].
//
// [Bloock Identity service]: https://bloock.com
type IdentityClient struct {
	bridgeClient bridge.BloockBridge
	configData   *proto.ConfigData
}

// NewIdentityClient creates a new instance of the IdentityClient with default configuration.
func NewIdentityClient() IdentityClient {
	return IdentityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   config.NewConfigDataDefault(),
	}
}

// NewIdentityClientWithConfig creates a new instance of the IdentityClient with the provided configuration.
func NewIdentityClientWithConfig(configData *proto.ConfigData) IdentityClient {
	return IdentityClient{
		bridgeClient: bridge.NewBloockBridge(),
		configData:   configData,
	}
}

// CreateIdentity creates a new identity.
func (c *IdentityClient) CreateIdentity(issuerKey identityV2.IdentityKey, params identityV2.DidParams) (string, error) {
	res, err := c.bridgeClient.IdentityV2().CreateIdentity(context.Background(), &proto.CreateIdentityV2Request{
		IssuerKey:  issuerKey.ToProto(),
		DidParams:  identityV2.DidParamsToProto(params),
		ConfigData: c.configData,
	})

	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return res.Did, nil
}

// CreateIssuer creates a new issuer on the Bloock Identity service.
func (c *IdentityClient) CreateIssuer(issuerKey identityV2.IdentityKey, publishInterval identityV2.PublishIntervalParams, params identityV2.DidParams, name, description, image string) (string, error) {
	var iName, iDescription, iImage *string
	if name != "" {
		iName = &name
	}
	if description != "" {
		iDescription = &description
	}
	if image != "" {
		iImage = &image
	}

	res, err := c.bridgeClient.IdentityV2().CreateIssuer(context.Background(), &proto.CreateIssuerRequest{
		IssuerKey:       issuerKey.ToProto(),
		IssuerParams:    identityV2.DidParamsToProto(params),
		Name:            iName,
		Description:     iDescription,
		Image:           iImage,
		PublishInterval: identityV2.PublishIntervalParamsToProto[publishInterval],
		ConfigData:      c.configData,
	})

	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return res.Did, nil
}

// GetIssuerByKey retrieves the DID of an issuer based on the issuer key and DID parameters.
func (c *IdentityClient) GetIssuerByKey(issuerKey identityV2.IdentityKey, params identityV2.DidParams) (string, error) {
	res, err := c.bridgeClient.IdentityV2().GetIssuerByKey(context.Background(), &proto.GetIssuerByKeyRequest{
		ConfigData:   c.configData,
		IssuerKey:    issuerKey.ToProto(),
		IssuerParams: identityV2.DidParamsToProto(params),
	})
	if err != nil {
		return "", err
	}

	if res.Error != nil {
		return "", errors.New(res.Error.Message)
	}

	return res.GetDid(), nil
}

// BuildSchema creates a new schema builder for defining a schema on the Bloock Identity service.
func (c *IdentityClient) BuildSchema(displayName string, schemaType, version, description string) identityV2.SchemaBuilder {
	return identityV2.NewSchemaBuilder(displayName, schemaType, version, description, c.configData)
}

// GetSchema retrieves a schema from the Bloock Identity service based on the schema ID (ex: Qma1t4uzbnB93E4rasNdu5UWMDh5qg3wMkPm68cnEyfnoM).
func (c *IdentityClient) GetSchema(id string) (identityV2.Schema, error) {
	res, err := c.bridgeClient.IdentityV2().GetSchema(context.Background(), &proto.GetSchemaRequestV2{
		ConfigData: c.configData,
		Id:         id,
	})

	if err != nil {
		return identityV2.Schema{}, err
	}

	if res.Error != nil {
		return identityV2.Schema{}, errors.New(res.Error.Message)
	}

	return identityV2.NewSchemaFromProto(res.GetSchema()), nil
}

// BuildCredential creates a new credential builder for defining a credential on the Bloock Identity service.
func (c *IdentityClient) BuildCredential(schemaId, issuerDid, holderDid string, expiration int64, version int32) identityV2.CredentialBuilder {
	return identityV2.NewCredentialBuilder(schemaId, issuerDid, holderDid, expiration, version, c.configData)
}

// PublishIssuerState publishes the state of an issuer on the Bloock Identity service.
func (c *IdentityClient) PublishIssuerState(issuerDid string, signer authenticity.Signer) (identityV2.IssuerStateReceipt, error) {
	res, err := c.bridgeClient.IdentityV2().PublishIssuerState(context.Background(), &proto.PublishIssuerStateRequest{
		ConfigData: c.configData,
		IssuerDid:  issuerDid,
		Signer:     signer.ToProto(),
	})

	if err != nil {
		return identityV2.IssuerStateReceipt{}, err
	}

	if res.Error != nil {
		return identityV2.IssuerStateReceipt{}, errors.New(res.Error.Message)
	}

	return identityV2.NewIssuerStateReceiptFromProto(res.GetStateReceipt()), nil
}

// GetCredentialProof retrieves the proof of a credential on the Bloock Identity service.
func (c *IdentityClient) GetCredentialProof(issuerDid string, credentialId string) (identityV2.CredentialProof, error) {
	res, err := c.bridgeClient.IdentityV2().GetCredentialProof(context.Background(), &proto.GetCredentialProofRequest{
		ConfigData:   c.configData,
		IssuerDid:    issuerDid,
		CredentialId: credentialId,
	})

	if err != nil {
		return identityV2.CredentialProof{}, err
	}

	if res.Error != nil {
		return identityV2.CredentialProof{}, errors.New(res.Error.Message)
	}

	return identityV2.NewCredentialProofFromProto(res.GetProof()), nil
}

// RevokeCredential revokes a credential on the Bloock Identity service.
func (c *IdentityClient) RevokeCredential(credential identityV2.Credential, signer authenticity.Signer) (bool, error) {
	res, err := c.bridgeClient.IdentityV2().RevokeCredential(context.Background(), &proto.RevokeCredentialRequestV2{
		ConfigData: c.configData,
		Credential: credential.ToProto(),
		Signer:     signer.ToProto(),
	})

	if err != nil {
		return false, err
	}

	if res.Error != nil {
		return false, errors.New(res.Error.Message)
	}

	return res.Result.GetSuccess(), nil
}

// CreateVerification creates a new verification session on the identity managed API provided.
func (c *IdentityClient) CreateVerification(proofRequest string) (identityV2.VerificationReceipt, error) {
	res, err := c.bridgeClient.IdentityV2().CreateVerification(context.Background(), &proto.CreateVerificationRequest{
		ConfigData:   c.configData,
		ProofRequest: proofRequest,
	})

	if err != nil {
		return identityV2.VerificationReceipt{}, err
	}

	if res.Error != nil {
		return identityV2.VerificationReceipt{}, errors.New(res.Error.Message)
	}

	return identityV2.NewVerificationReceiptFromProto(res.GetResult()), nil
}

// WaitVerification waits for the completion of a verification session on the identity managed API provided.
func (c *IdentityClient) WaitVerification(sessionID int64, params identityV2.VerificationParams) (bool, error) {
	if params.Timeout == 0 {
		params.Timeout = int64(120000)
	}

	res, err := c.bridgeClient.IdentityV2().WaitVerification(context.Background(), &proto.WaitVerificationRequest{
		ConfigData: c.configData,
		SessionId:  sessionID,
		Timeout:    params.Timeout,
	})

	if err != nil {
		return false, err
	}

	if res.Error != nil {
		return false, errors.New(res.Error.Message)
	}

	return res.GetStatus(), nil
}

// GetVerificationStatus retrieves the status of a verification session on the identity managed API provided.
func (c *IdentityClient) GetVerificationStatus(sessionID int64) (bool, error) {
	res, err := c.bridgeClient.IdentityV2().GetVerificationStatus(context.Background(), &proto.GetVerificationStatusRequest{
		ConfigData: c.configData,
		SessionId:  sessionID,
	})

	if err != nil {
		return false, err
	}

	if res.Error != nil {
		return false, errors.New(res.Error.Message)
	}

	return res.GetStatus(), nil
}

File: languages/js/src/client/identity_v2.ts
import { BloockBridge } from "../bridge/bridge";
import { ConfigData } from "../bridge/proto/config";
import {
  CreateIssuerRequest,
  GetIssuerByKeyRequest,
  GetCredentialProofRequest,
  RevokeCredentialRequestV2,
  GetSchemaRequestV2,
  CreateIdentityV2Request,
  PublishIssuerStateRequest,
  CreateVerificationRequest,
  WaitVerificationRequest,
  GetVerificationStatusRequest
} from "../bridge/proto/identity_v2";
import { NewConfigData } from "../config/config";
import {
  IssuerStateReceipt,
  PublishIntervalParams,
  Schema,
  VerificationReceipt
} from "../entity/identity_v2";
import { Credential } from "../entity/identity_v2/credential";
import { CredentialBuilder } from "../entity/identity_v2/credential_builder";
import { CredentialProof } from "../entity/identity_v2/credential_proof";
import { IdentityKey } from "../entity/identity_v2/identity_key";
import { DidParams } from "../entity/identity_v2/did_params";
import { SchemaBuilder } from "../entity/identity_v2/schema_builder";
import { Signer } from "../entity/authenticity";

/**
 * Represents a client for interacting with the [Bloock Identity service](https://dashboard.bloock.com/login).
 */
export class IdentityClient {
  private bridge: BloockBridge;
  private configData: ConfigData;

  /**
   * Creates a new instance of the IdentityClient with default configuration.
   * @param configData 
   */
  constructor(configData?: ConfigData) {
    this.bridge = new BloockBridge();
    this.configData = NewConfigData(configData);
  }

  /**
   * Creates a new identity.
   * @param issuerKey 
   * @param didParams 
   * @returns 
   */
  public createIdentity(
    issuerKey: IdentityKey,
    didParams?: DidParams
  ): Promise<string> {
    const request = CreateIdentityV2Request.fromPartial({
      issuerKey: issuerKey.toProto(),
      didParams: didParams?.toProto(),
      configData: this.configData
    });

    return this.bridge
      .getIdentityV2()
      .CreateIdentity(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.did;
      });
  }

  /**
   * Creates a new issuer on the Bloock Identity service.
   * @param issuerKey 
   * @param publishInterval 
   * @param issuerParams 
   * @param name 
   * @param description 
   * @param image 
   * @returns 
   */
  public createIssuer(
    issuerKey: IdentityKey,
    publishInterval: PublishIntervalParams,
    issuerParams?: DidParams,
    name?: string,
    description?: string,
    image?: string
  ): Promise<string> {
    const request = CreateIssuerRequest.fromPartial({
      issuerKey: issuerKey.toProto(),
      issuerParams: issuerParams?.toProto(),
      name: name,
      description: description,
      image: image,
      publishInterval: PublishIntervalParams.toProto(publishInterval),
      configData: this.configData
    });

    return this.bridge
      .getIdentityV2()
      .CreateIssuer(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.did;
      });
  }

  /**
   * Gets the DID of an issuer based on the issuer key.
   * @param issuerKey 
   * @param issuerParams 
   * @returns 
   */
  public getIssuerByKey(
    issuerKey: IdentityKey,
    issuerParams?: DidParams
  ): Promise<string> {
    const request = GetIssuerByKeyRequest.fromPartial({
      issuerKey: issuerKey.toProto(),
      issuerParams: issuerParams?.toProto(),
      configData: this.configData
    });

    return this.bridge
      .getIdentityV2()
      .GetIssuerByKey(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.did;
      });
  }

  /**
   * Creates a new schema builder for defining a schema on the Bloock Identity service.
   * @param displayName 
   * @param schemaType 
   * @param version 
   * @param description 
   * @returns 
   */
  public buildSchema(
    displayName: string,
    schemaType: string,
    version: string,
    description: string
  ): SchemaBuilder {
    return new SchemaBuilder(
      displayName,
      schemaType,
      version,
      description,
      this.configData
    );
  }

  /**
   * Gets a schema from the Bloock Identity service based on the schema ID (ex: Qma1t4uzbnB93E4rasNdu5UWMDh5qg3wMkPm68cnEyfnoM).
   * @param id 
   * @returns 
   */
  public getSchema(id: string): Promise<Schema> {
    const request = GetSchemaRequestV2.fromPartial({
      configData: this.configData,
      id: id
    });

    return this.bridge
      .getIdentityV2()
      .GetSchema(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return Schema.fromProto(res.schema!);
      });
  }

  /**
   * Creates a new credential builder for defining a credential on the Bloock Identity service.
   * @param schemaId 
   * @param issuerDid 
   * @param holderDid 
   * @param expiration 
   * @param version 
   * @returns 
   */
  public buildCredential(
    schemaId: string,
    issuerDid: string,
    holderDid: string,
    expiration: number,
    version: number
  ): CredentialBuilder {
    return new CredentialBuilder(
      schemaId,
      issuerDid,
      holderDid,
      expiration,
      version,
      this.configData
    );
  }

  /**
   * Publishes the state of an issuer on the Bloock Identity service.
   * @param issuerDid 
   * @param signer 
   * @returns 
   */
  public publishIssuerState(
    issuerDid: string,
    signer: Signer
  ): Promise<IssuerStateReceipt> {
    const req = PublishIssuerStateRequest.fromPartial({
      configData: this.configData,
      issuerDid: issuerDid,
      signer: signer.toProto()
    });

    return this.bridge
      .getIdentityV2()
      .PublishIssuerState(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return IssuerStateReceipt.fromProto(res.stateReceipt!);
      });
  }

  /**
   * Gets the proof of a credential on the Bloock Identity service.
   * @param issuerDid 
   * @param credentialId 
   * @returns 
   */
  public getCredentialProof(
    issuerDid: string,
    credentialId: string
  ): Promise<CredentialProof> {
    const request = GetCredentialProofRequest.fromPartial({
      issuerDid: issuerDid,
      credentialId: credentialId,
      configData: this.configData
    });

    return this.bridge
      .getIdentityV2()
      .GetCredentialProof(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return CredentialProof.fromProto(res.proof!);
      });
  }

  /**
   * Revokes a credential on the Bloock Identity service.
   * @param credential 
   * @param signer 
   * @returns 
   */
  public revokeCredential(
    credential: Credential,
    signer: Signer
  ): Promise<boolean> {
    const request = RevokeCredentialRequestV2.fromPartial({
      configData: this.configData,
      credential: credential.toProto(),
      signer: signer.toProto()
    });

    return this.bridge
      .getIdentityV2()
      .RevokeCredential(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.result!.success!;
      });
  }

  /**
   * Creates a new verification session on the identity managed API provided.
   * @param proofRequest 
   * @returns 
   */
  public createVerification(
    proofRequest: string
  ): Promise<VerificationReceipt> {
    const request = CreateVerificationRequest.fromPartial({
      configData: this.configData,
      proofRequest: proofRequest
    });

    return this.bridge
      .getIdentityV2()
      .CreateVerification(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return VerificationReceipt.fromProto(res.result!);
      });
  }

  /**
   * Waits for the completion of a verification session on the identity managed API provided.
   * @param sessionID 
   * @param timeout 
   * @returns 
   */
  public waitVerification(
    sessionID: number,
    timeout?: number
  ): Promise<boolean> {
    const request = WaitVerificationRequest.fromPartial({
      configData: this.configData,
      sessionId: sessionID,
      timeout: timeout !== null ? timeout : 120000
    });

    return this.bridge
      .getIdentityV2()
      .WaitVerification(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.status;
      });
  }

  /**
   * Gets the status of a verification session on the identity managed API provided.
   * @param sessionID 
   * @returns 
   */
  public getVerificationStatus(sessionID: number): Promise<boolean> {
    const request = GetVerificationStatusRequest.fromPartial({
      configData: this.configData,
      sessionId: sessionID
    });

    return this.bridge
      .getIdentityV2()
      .GetVerificationStatus(request)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return res.status;
      });
  }
}
